Enable i686 as a Host Platform

Kaiede opened the pull request:
<!-- What's in this pull request? -->
This is a set of changes that is aimed at enabling Linux i686 as a host platform for Swift. The reason behind this is to make it possible to use i686 as an analogue for other host platforms (Arm32), making investigation into 32-bit specific issues easier by being able to do the work on more powerful hardware where builds can easily be a quarter of the time. 

As a side effect, once other issues are also addressed, it would enable a new class of hardware to be used for Swift. It would also enable something like Debian and/or Ubuntu on 32-bit x86 to be a CI analogue for ARM32 Single Board Computers, which are popular, cheap, but very slow to build and report results. 

Each change is split to make the set easier to understand, and any comments about why a particular change was made the way it was should be available in the description of that change. 

The two changes marked with "[32-bit Linux]" also affect ARM 32 hosts (Raspberry Pi), and are regressions that happened after 4.1, and currently impact 4.2 and master. 

**Remaining Issues**

- Building the stdlib on i686 relies on PR #19600, which affects ARM 32 as well.
- Full support of building at the tip of development is still broken because of a recent change. In particular, the SILLocation static_assert recently added fires on 32-bit hosts at the moment. 

**Issues that are Out of Scope**

- There are still breaks in downstream projects (libdispatch, llbuild, etc). 
- LLVM doesn't know how to dynamically link certain ELF32 relocations: GOTOFF/GOT32 when operating as VM/JIT instead of a back-end. This will need to be fixed before the Swift PM will build/work. 


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Kaiede committed:
[i686 Linux] Bootstrap the build

Prior to this, the swift build didn’t understand what i686 is, or what to do about building it. This unblocks building, but will still run into build breaks.

Kaiede committed:
[i686 Linux] Fix undefined ‘__swift_ssize_t’

Kaiede committed:
[32-bit Linux] Align _GenericContext to 8 bytes

A static assert fires around GenericContext not quite being the correct size for 32-bit. Aligning _GenericContext addresses this problem.

swift/include/swift/AST/Decl.h:
 1541  static_assert(sizeof(_GenericContext) + sizeof(DeclContext) ==
 1542:               sizeof(GenericContext), "Please add fields to _GenericContext");

Kaiede committed:
[32-bit Linux] Handle size_t/uint64_t assumptions

There’s a few places where size_t is used for a field/parameter when constructing an array for types. Unfortunately, the Bitfields that were backing the inputs to these at some point after 4.1 grew past 32 bits and are now backed by a uint64_t. Even though the slice of the bitfield is small enough for 32-bit, clang sees these slices as 64-bit and complains if there isn’t a cast involved.

Kaiede committed:
[i686 Linux] Support Float80

Darwin and Linux use different size and alignments for “long double” on 32-bit x86. Instead of hardcoding it, we should just ask LLVM about it.

Kaiede committed:
[i686 Linux] Add __mulodi4 Math Stub to i686

Win and Linux Arm32 have the same aversion to linking compiler-rt for this.

jrose-apple reviewed (commented):
Compiler-side changes look good. @stephentyrone and @mikeash can check the stdlib parts, and @Rostepher, @zisko, and @compnerd are probably good for the CMake and build system parts.

mikeash commented:
Stdlib bits look OK. Not a whole lot to see there. The rest looked good too as far as I could tell, but I don't know those areas as well.

Rostepher commented:
The build system changes look sensible to me.

jrose-apple commented:
@swift-ci Please test

Kaiede committed:
[32-bit Linux] Adjust SILLocation Size Assert

The assert currently assumes that the size of SILLocation is 3 pointers big, or 24-bytes as it is on 64-bit platforms. The catch is that this math works because it’s having to make a rough assumption of the size of unsigned and pointer in relation to each other. 4 unsigned + 1 pointer = 3 pointer in this example.

This breaks down on 32-bit where 4 unsigned + 1 pointer = 5 unsigned = 20 bytes. It technically doesn’t violate the spirit of the assert, but the way the assert is written simply doesn’t work for 32-bit.

It can be written as `sizeof(SILLocation) <= 6*sizeof(unsigned)` without an ifdef, but it means that if someone was to make it even smaller at some point, the assert wouldn’t fail, signaling to the developer that they should update the assert with the new, smaller, value. But it would be cleaner. The way this change is written, if the size changes in either direction, the developer is signaled that they should either update the assert, or fix their regression.

Kaiede commented:
@jrose-apple Sorry, but can you take a look at the SILLocation assert change? I didn't get a good chance to see what was up with it, and I'm not 100% confident this is the approach you'd want here. But it would be nice to get that fix in at the same time. That way, this PR and #19600 are effectively everything needed to get master building on i686, at least as of when I pulled into my fork yesterday.

Kaiede commented on lib/SIL/SILLocation.cpp:
```diff
@@ -20,8 +20,14 @@
 using namespace swift;
 
+// 64-bit is 24 bytes, 32-bit is 20 bytes.
+#ifdef __LP64__
```
I'll go ahead and make that change for the ifdef tomorrow (Monday). I think the only way you can deflate this type any more is to remove stuff out of the union so it can shrink. There's no padding on 32-bit x86 at least, and ARM shouldn't be any different.

I think the only concern with the macro is that at least as of a few days ago, it wasn't used anywhere in the swift project. So I guess it's more of a question of: is this where swift wants to go all-up as their macro for checking for 64-bit in the future? I personally think being able to find the various places 64-bit is explicitly checked for is valuable, so I'm a little hesitant to fracture that. 

But I like the general idea.

jrose-apple replied:
I looked around and LP64 isn't really being used either. I think we just haven't had anyone really put effort into building a 32-bit version of the compiler.

(In the runtime we use `__POINTER_WIDTH__ == 64`, but that's because the runtime is built with the same Clang that's embedded in the Swift compiler.)

Kaiede replied:
I agree it isn't used in many places, but 8 is still not zero. I agree with the suggestion, I just want to make sure this is also a statement of intent for the future too, is all.

----

What got me looking at this stuff in the first place is that there are 3.1.1 binaries for the Raspberry Pi floating around that work quite well and are stable. But with folks moving to support only 4.x, it's making it hard to stick with 3.1.1. 

4.1 has even been mostly working so long as you didn't use Foundation for anything, because of a nasty bug lurking there (we've got the community binaries unblocked, but still need to re-fix the master branch itself).

I'm mostly trying to get things into a state where the community isn't facing the daunting prospect of dealing with regressions every major release, which lags adoption of new versions by half a year or more. Mostly because I think Swift is working really well on these single-board computers as an alternative to Node or Python. It just really needs some stability, and this is a step towards that.

jrose-apple replied:
Again, you have to distinguish the uses in the runtime and tests (target conditions) from the uses in the compiler (host conditions). The only one that's in the compiler is the one that's `defined(__BIG_ENDIAN__) && defined(__LP64__)`, and I'm not even sure that condition is correct anyway.

Kaiede replied:
That's fair, although I'd imagine you'd probably not want too much LP64 leakage there either?

That big endian condition is "correct" in the sense that you have to get pretty niche (more so than the single board computers) to make it fail.

jrose-apple replied:
I think my concern is that that should have been a test of the target platform as well, rather than the host. If anyone is cross-compiling from a big-endian 64-bit compiler to a 32-bit anything, something funny's going to happen here.

Kaiede replied:
I've redone the assert, this time just being more explicit about the underlying types. The #ifdef also works, but this has the same effect without an ifdef. It's also a bit easier to understand how the byte size was reached and correlate it with the SILLocation class' structure.

jrose-apple replied:
All right, that works. Thanks!

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

compnerd commented:
LGTM
